packages/apollo-cli/src/Config.ts

```diff
@@ -0,0 +1,184 @@
+import fs from 'node:fs'
+
+import Joi from 'joi'
+import YAML from 'yaml'
+
+import { ConfigError } from './utils.ts'
+
+interface RootCredentials {
+  username?: string
+  password?: string
+}
+
+interface Profile {
+  address?: string
+  accessType?: string
+  accessToken?: string
+  rootCredentials?: RootCredentials
+}
+
+const KEYS = [
+  'address',
+  'accessType',
+  'accessToken',
+  'rootCredentials.username',
+  'rootCredentials.password',
+]
+
+interface RecursiveObject {
+  [key: number | string]: RecursiveObject | unknown
+}
+
+function isValidAddress(address: string): boolean {
+  const port: string | undefined = address.split(':').pop()
+  if (port === undefined || /^\d+$/.test(port) === false) {
+    return false
+  }
+
+  let url
+  try {
+    url = new URL(address)
+  } catch {
+    return false
+  }
+
+  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
+    return false
+  }
+  return true
+}
+
+const validateAddress = (address: string, helpers: any) => {
+  if (isValidAddress(address)) {
+    return address
+  }
+  return helpers.error('any.invalid')
+}
+
+const profileSchema = Joi.object({
+  address: Joi.string().custom(validateAddress),
+  accessType: Joi.string().valid('google', 'microsoft', 'root'),
+  accessToken: Joi.string(),
+  rootCredentials: {
+    username: Joi.string().allow(null, ''),
+    password: Joi.string().allow(null, ''),
+  },
+})
+const configSchema = Joi.object().pattern(/.*/, profileSchema)
```

Comment:
This is related to the above comment on the Profile type, I think this schema captures the same thing we're trying to describe in the updated type.

```suggestion
const profileSchema = Joi.object({
  address: Joi.string()
    .uri({ scheme: /https?/ })
    .required(),
  accessType: Joi.string()
    .valid('google', 'microsoft', 'root', 'guest')
    .required(),
  accessToken: Joi.string(),
  rootCredentials: Joi.object({
    username: Joi.string().required(),
    password: Joi.string().required(),
  }).when('accessType', {
    is: Joi.string().valid('root'),
    // eslint-disable-next-line unicorn/no-thenable
    then: Joi.required(),
    otherwise: Joi.forbidden(),
  }),
})
const configSchema = Joi.object().pattern(Joi.string(), profileSchema)
```